client/apis/volumegroupsnapshot/v1alpha1/types.go

```diff
@@ -0,0 +1,328 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// +kubebuilder:object:generate=true
+package v1alpha1
+
+import (
+	core_v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
+)
+
+// VolumeGroupSnapshotSpec defines the desired state of VolumeGroupSnapshot
+type VolumeGroupSnapshotSpec struct {
+	// Selector is a label query over PersistentVolumeClaims that are to be grouped
+	// together for snapshotting.
+	// This labelSelector will be used to match the label added to a PVC.
+	// If the label is added or removed to a volume after a group snapshot
+	// is created, the existing group snapshots won't be modified.
+	// Once a VolumeGroupSnapshotContent is created and the sidecar starts to process
+	// it, the volume list will not change with retries.
+	// Required.
+	Selector metav1.LabelSelector `json:"selector" protobuf:"bytes,1,opt,name=selector"`
```

Comment:
I think something is missing from the KEP.  We need the following for pre-provisioned VolumeGroupSnapshot to work.

```
type VolumeGroupSnapshotSpec struct {
	// source specifies where a group snapshot will be created from.
	// This field is immutable after creation.
	// Required.
	Source VolumeGroupSnapshotSource `json:"source" protobuf:"bytes,1,opt,name=source"`

	// VolumeGroupSnapshotClassName is the name of the VolumeGroupSnapshotClass
	// requested by the VolumeGroupSnapshot.
	// VolumeGroupSnapshotClassName may be left nil to indicate that the default
	// class will be used.
	// Empty string is not allowed for this field.
	// +optional
	VolumeGroupSnapshotClassName *string `json:"volumeGroupSnapshotClassName,omitempty" protobuf:"bytes,2,opt,name=volumeGroupSnapshotClassName"`
}


// VolumeGroupSnapshotSource specifies whether the underlying group snapshot should be
// dynamically taken upon creation or if a pre-existing VolumeGroupSnapshotContent
// object should be used.
// Exactly one of its members must be set.
// Members in VolumeGroupSnapshotSource are immutable.
type VolumeGroupSnapshotSource struct {
	// together for snapshotting.
	// This labelSelector will be used to match the label added to a PVC.
	// If the label is added or removed to a volume after a group snapshot
	// is created, the existing group snapshots won't be modified.
	// Once a VolumeGroupSnapshotContent is created and the sidecar starts to process
	// it, the volume list will not change with retries.
	// This field is immutable.
	// +optional
	Selector metav1.LabelSelector `json:"selector" protobuf:"bytes,1,opt,name=selector"`
	
	// volumeGroupSnapshotContentName specifies the name of a pre-existing VolumeGroupSnapshotContent
	// object representing an existing volume group snapshot.
	// This field should be set if the volume group snapshot already exists and only needs a representation in Kubernetes.
	// This field is immutable.
	// +optional
	VolumeGroupSnapshotContentName *string `json:"volumeGroupSnapshotContentName,omitempty" protobuf:"bytes,2,opt,name=volumeGroupSnapshotContentName"`
}

```

Reply:
Ah I thought it looked a little strange. I've updated the spec